<div align="center">
    <h1> My Machine Learning Journey</h1>
Civil Engineer · Researcher Youtuber · Machine Learning Engineer (The Goal)
    <br>
    Sharing and documenting my progress and journey as I learn machine learning
     <br>
</div>

<br>

<div align="center">
    <a target="_blank" href="https://github.com/barbaraaboagye/My-MachineLearning-Journey"><img src="https://img.shields.io/github/last-commit/barbaraaboagye/My-MachineLearning-Journey"></a>&nbsp;
      <a target="_blank" href="https://www.youtube.com/channel/UCEYKFq7ZEg81GYxpzNqYZ4)"><img src="https://img.shields.io/youtube/channel/subscribers/UCEYKFq7ZEg81GYxpzNqYZ4Q?style=social"></a>&nbsp;
    <a target="_blank" href="https://fr.linkedin.com/in/barbara-aboagye-233ba8133"><img src="https://img.shields.io/badge/style--5eba00.svg?label=LinkedIn&logo=linkedin&style=social"></a>&nbsp;
    <a target="_blank" href="https://twitter.com/awesome_ama"><img src="https://img.shields.io/twitter/follow/awesome_ama?style=social"></a>
    <br>
</div>

<br>

# Slicing, Joining and Gradients - Day x of x

Hours completed : 1hr 30min <br>


# Resources used : 

- [Foundation- PyTorch](https://madewithml.com/courses/foundations/pytorch/)
<br>

Handwritten notes can be found [here][def]

[def]: https://github.com/barbaraaboagye/My-MachineLearning-Journey/blob/a4c272eca2f897b2d0cc32afab1afa5f58d739f2/Handwritten%20notes/230403_SlicingJoiningGradients_PytorchFoundations.pdf

### Set up 

In [37]:
import numpy as np
import torch

In [38]:
SEED = 1234

In [39]:
# Set seed for reproducibility

np.random.seed(seed=SEED)
torch.manual_seed(SEED)

### Basics 

We'll first cover some basics with PyTorch such as creating tensors and converting from common data structures (lists, arrays, etc.) to tensors.

In [40]:
# Creating a random tensor

x = torch.randn(2,3) # normal distribution (rand(2,3) -> uniform distribution)
print (f"Type: {x.type()}")
print (f"Size: {x.shape}")
print (f"Values: \n{x}")

Type: torch.FloatTensor
Size: torch.Size([2, 3])
Values: 
tensor([[ 0.0461,  0.4024, -1.0115],
        [ 0.2167, -0.6123,  0.5036]])


This code creates a tensor `x` of size `(2,3)` filled with random values from a normal distribution with mean 0 and standard deviation 1 using PyTorch.

The tensor `x` has 2 rows and 3 columns

In [41]:
#Zeros and Ones tensor

x = torch.zeros(2, 3)
print(x)
x = torch.ones(2,3)
print(x)

tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])


In [42]:
# List to Tensor

x = [1,2,3], [4,5,6 ] # normal list
print (x)
print(np.shape(x))
print(f"Type: {np.ndim(x)}")
y = torch.Tensor([[1,2,3], [4,5,6 ]])
print (f"Size: {y.shape}")
print(f"x:\n {y}")
z = torch.Tensor(x)
print (f"Size: {z.shape}")
print(f"x:\n {z}")

([1, 2, 3], [4, 5, 6])
(2, 3)
Type: 2
Size: torch.Size([2, 3])
x:
 tensor([[1., 2., 3.],
        [4., 5., 6.]])
Size: torch.Size([2, 3])
x:
 tensor([[1., 2., 3.],
        [4., 5., 6.]])


In [43]:
## NumPy array to Tensor

x = torch.Tensor(np.random.rand(2,3))
print (f"Size: {x.shape}")
print(f"x:\n {x}")

Size: torch.Size([2, 3])
x:
 tensor([[0.1915, 0.6221, 0.4377],
        [0.7854, 0.7800, 0.2726]])


In [44]:
# Changing tensor type 
x = torch.Tensor(3,4)
print(f"Type: {x.type()}")
print(x)
x = x.long()
print(f"Type: {x.type()}")
print(x)

Type: torch.FloatTensor
tensor([[3.4332e-06, 1.0527e-11, 6.6651e-10, 2.1346e-07],
        [5.4654e+22, 6.4591e-07, 1.3120e-08, 5.2713e-08],
        [4.1652e-11, 7.1450e+31, 4.1418e-41, 0.0000e+00]])
Type: torch.LongTensor
tensor([[                   0,                    0,                    0,
                            0],
        [-9223372036854775808,                    0,                    0,
                            0],
        [                   0, -9223372036854775808,                    0,
                            0]])


### Long tensor 

In programming, we often work with numbers, like 1, 2, 3, etc. However, sometimes we need to work with really large numbers that are too big to be represented using the normal integer data type.

For example, imagine you have a dataset with billions of entries, and you need to index into it using unique identifiers that are assigned to each entry. In this case, you would need a data type that can represent really large integer values, and that's where LongTensor comes in.

In PyTorch, a LongTensor is a special type of data structure that can hold large integer values ranging from -9223372036854775808 to 9223372036854775807. You can think of it like a container that can store really big numbers.

You can create a LongTensor by either explicitly converting an existing tensor of a different data type using the .long() method or by directly initializing it using the torch.LongTensor() constructor. Once you have a LongTensor, you can use it for various mathematical operations, just like any other tensor in PyTorch.

#### Operations

In [45]:
# Addition

x = torch.randn(2,3)
y = torch.randn(2,3)
z = x + y
print(f"Size: {z.shape}")
print(f"Values: \n{z}")

Size: torch.Size([2, 3])
Values: 
tensor([[ 0.0761, -0.6775, -0.3988],
        [ 3.0633, -0.1589,  0.3514]])


In [46]:
## Dot product

x = torch.randn(2,3)
y = torch.randn(3,2)
z = torch.mm(x,y)
print(f"Size : {z.shape}")
print(f"Values : \n {z}")

Size : torch.Size([2, 2])
Values : 
 tensor([[ 1.0796, -0.0759],
        [ 1.2746, -0.5134]])


The `torch.mm()` function is used to perform matrix multiplication between two tensors of size `(m,n)` and `(n,p)`, resulting in a tensor of size `(m,p)`. In other words, it multiplies each element of each row of the first tensor with each element of each column of the second tensor, and then sums the products to get the corresponding element of the resulting tensor.

In [47]:
# Transpose

x = torch.randn(2,3)
print(f"Size : {x.shape}")
print (f"Values : \n{x}")
y = torch.t(x)
print(f"Size: {y.shape}")
print(f"Values: \n{y}")

Size : torch.Size([2, 3])
Values : 
tensor([[ 0.8042, -0.1383,  0.3196],
        [-1.0187, -1.3147,  2.5228]])
Size: torch.Size([3, 2])
Values: 
tensor([[ 0.8042, -1.0187],
        [-0.1383, -1.3147],
        [ 0.3196,  2.5228]])


In [48]:
## Reshape 
x = torch.randn(2,3)
print(f"Values x:\n {x}")
z = x.view(3,2)
print(f"Size: {z.shape}")
print (f"Values z:\n{z}")

Values x:
 tensor([[ 0.4501,  0.2709, -0.8087],
        [-0.0217, -1.0413,  0.0702]])
Size: torch.Size([3, 2])
Values z:
tensor([[ 0.4501,  0.2709],
        [-0.8087, -0.0217],
        [-1.0413,  0.0702]])


This code reshapes the tensor `x` using the `.view()` method in PyTorch and assigns the resulting tensor to `z`.

The `.view()` method in PyTorch is used to reshape a tensor to a new shape while maintaining the total number of elements in the tensor. In this case, the tensor `x` has dimensions `(2,3)` and a total of 6 elements.

The `.view(3, 2)` method is used to reshape `x` into a tensor with dimensions `(3,2)` which also has a total of 6 elements. The resulting tensor `z` will have 3 rows and 2 columns.

Note that when reshaping a tensor, the order of the elements in memory is preserved. In other words, the elements are simply rearranged to form the new shape without any change in their values. If the new shape does not have the same number of elements as the original tensor, a runtime error will be raised.

In [49]:
# Dangers of reshaping (unintended consequences)
x = torch.tensor([
    [[1,1,1,1], [2,2,2,2], [3,3,3,3]],
    [[10,10,10,10], [20,20,20,20], [30,30,30,30]]
])
print(f"Size: {x.shape}")
print(f"x: \n{x}\n")

a = x.view(x.size(1), -1)
print(f"\nSize: {a.shape}")
print(f"a: \n{a}\n")

b = x.transpose(0,1).contiguous()
print(f"\nSize: {b.shape}")
print(f"b: \n{b}\n")

c = b.view(b.size(0), -1)
print(f"\nSize: {c.shape}")
print(f"c: \n{c}")


Size: torch.Size([2, 3, 4])
x: 
tensor([[[ 1,  1,  1,  1],
         [ 2,  2,  2,  2],
         [ 3,  3,  3,  3]],

        [[10, 10, 10, 10],
         [20, 20, 20, 20],
         [30, 30, 30, 30]]])


Size: torch.Size([3, 8])
a: 
tensor([[ 1,  1,  1,  1,  2,  2,  2,  2],
        [ 3,  3,  3,  3, 10, 10, 10, 10],
        [20, 20, 20, 20, 30, 30, 30, 30]])


Size: torch.Size([3, 2, 4])
b: 
tensor([[[ 1,  1,  1,  1],
         [10, 10, 10, 10]],

        [[ 2,  2,  2,  2],
         [20, 20, 20, 20]],

        [[ 3,  3,  3,  3],
         [30, 30, 30, 30]]])


Size: torch.Size([3, 8])
c: 
tensor([[ 1,  1,  1,  1, 10, 10, 10, 10],
        [ 2,  2,  2,  2, 20, 20, 20, 20],
        [ 3,  3,  3,  3, 30, 30, 30, 30]])


When we reshape a tensor, we need to be careful. Reshaping can change the order of the elements in the tensor, which can cause problems if we're not expecting it. In the code above, we first create a three-dimensional tensor `x`. Then, we reshape it into a tensor a with dimensions `(3, 8)`. This flattens the tensor, so that all the elements are in a single row.

Next, we transpose the first two dimensions of `x` to get a tensor `b`. However, `b` is not contiguous, which means that its elements are not in a simple, ordered sequence in memory. This can cause issues later on.

Finally, we reshape `b` into a tensor c with dimensions `(2, 12)` using the .view() method. This could also cause issues, since the elements of b are not contiguous.

So, when we reshape tensors, we need to be careful and make sure we understand the order of the element

In [50]:
## Dimensional operations
x = torch.randn(2, 3)
print(f"Values: \n{x}")
y = torch.sum(x, dim=0) # add each row's value for every column
print(f"Values: \n{y}")
z = torch.sum(x, dim=1) # add each columns's value for every row
print(f"Values: \n{z}")


Values: 
tensor([[ 0.5797, -0.0599,  0.1816],
        [-0.6797, -0.2567, -1.8189]])
Values: 
tensor([-0.1000, -0.3166, -1.6373])
Values: 
tensor([ 0.7013, -2.7553])


### Indexing 


In [51]:
x = torch.randn(3,4)
print (f"x: \n{x}")
print (f"x[:1] : \n{x[:1]} ") # select first row
print (f"x[:1,1:3] \n {x[:1,1:3]} ")

x: 
tensor([[ 0.2111,  0.3372,  0.6638,  1.0397],
        [ 1.8434,  0.6588, -0.2349, -0.0306],
        [ 1.7462, -0.0722, -1.6794, -1.7010]])
x[:1] : 
tensor([[0.2111, 0.3372, 0.6638, 1.0397]]) 
x[:1,1:3] 
 tensor([[0.3372, 0.6638]]) 


The second print statement displays a slice of the tensor `x` using indexing. Specifically, it uses the slice notation `x[:1]` to select the first row of the tensor.

The syntax `x[:1, 1:3]` means "select the first row (`:1`) and columns 1 to 2 (`1:3`) of the tensor `x`". The resulting sub-tensor will have shape `(1, 2)`, since it has one row and two columns.

### Slicing

Slicing is a way to extract a portion of a sequence or an array in a specific order. 

In PyTorch, slicing is performed using the square bracket notation (`[]`). You can use slicing to extract a subset of elements from a tensor by specifying the starting and ending indices of the slice for each dimension. For example, `x[:, 2:4]` selects all rows and columns 2 to 3 of tensor `x`.

Slicing can also be used to modify a tensor by assigning a new value to a slice of the tensor. For example, `x[1,:] = torch.zeros(3)` sets the second row of tensor `x` to all zeros.

Slicing is a powerful technique that allows you to efficiently extract and manipulate subsets of data in tensors. It is commonly used in data processing, machine learning, and deep learning applications.

In [52]:
# Select with dimensional indices

x = torch.randn(2,3)
print(f"Values x: \n{x}")

col_indices = torch.LongTensor([0,2])
chosen = torch.index_select(x,dim = 1,index = col_indices) #values from column
print(f"Values : \n{chosen}")

row_indices = torch.LongTensor([0,1])
col_indices = torch.LongTensor([0,2])
chosen = x[row_indices,col_indices]
print(f" Values : \n{ chosen}")

Values x: 
tensor([[ 0.6486,  1.7653,  1.0812],
        [ 1.2436,  0.8971, -0.0784]])
Values : 
tensor([[ 0.6486,  1.0812],
        [ 1.2436, -0.0784]])
 Values : 
tensor([ 0.6486, -0.0784])


In this code, a 2x3 tensor `x` is created using the `torch.randn()` method, which generates random numbers from a normal distribution with mean 0 and standard deviation 1. Then, two examples of selecting values from `x` using dimensional indices are demonstrated.

The first example selects values from specific columns of the tensor `x`. A tensor `col_indices` is created that contains the indices of the columns to select (in this case, columns 0 and 2). The `torch.index_select()` method is used to select the specified columns of `x`. The `dim` parameter specifies the dimension along which to select values (in this case, dimension 1, corresponding to columns). The resulting tensor `chosen` will have shape `(2, 2)` since it has two rows and two columns.

The second example selects values from specific rows and columns of the tensor `x`. Two tensors, `row_indices` and `col_indices`, are created that contain the indices of the rows and columns to select (in this case, row 0 column 0, and row 1 column 2). The indices are used to index into `x` directly using square brackets. The resulting tensor `chosen` will have shape `(2,)` since it has two elements.

### Why are we indexing with `LongTensor`?

In PyTorch, tensors can have different data types such as `FloatTensor`, `DoubleTensor`, `LongTensor`, `IntTensor`, and so on. The data type of a tensor determines the precision of the values that it can hold and the operations that can be performed on it.

In the example you provided, `col_indices` is a tensor that stores the column indices that we want to select from `x`. We use `torch.LongTensor` to create `col_indices` because it is an integer tensor.

By default, PyTorch creates `FloatTensor` when you pass a list of floating-point values to the constructor, and it creates `LongTensor` when you pass a list of integers. If we had used `torch.Tensor([0, 2])` instead of `torch.LongTensor([0, 2])`, PyTorch would have created a `FloatTensor` with elements `0.0` and `2.0`, which would not be valid indices for `x`.

Therefore, it is important to use the correct data type for the indices when selecting elements from a tensor. In this case, we use `torch.LongTensor` to ensure that the indices are valid and that the tensor indexing operation works as expected

### Will using `IntTensor` work?

Yes, using `IntTensor` will also work in most cases where `LongTensor` is used.

In PyTorch, `IntTensor` and `LongTensor` are both integer data types that can be used to store and manipulate integer values. The main difference between them is their size. `IntTensor` is a 32-bit integer data type, while `LongTensor` is a 64-bit integer data type.

For most applications, using `IntTensor` instead of `LongTensor` will work fine. However, if you are working with very large tensors or very large indices, you may need to use `LongTensor` to avoid overflow errors.

In general, it is a good practice to use the correct data type for the specific use case. For example, if you are working with very large tensors or very large indices, you should use `LongTensor`. If you are working with smaller tensors or indices, you can use `IntTensor`.


### Joining

Tensors can be combined via **concatenation** or **stacking** operation, which are similar with [Numpy's joining operation](https://github.com/barbaraaboagye/My-MachineLearning-Journey/blob/052b4aa4d52db40e2ed5e4814943aa3849299ad2/Lessons/230323_TransposeReshapeJoinExpand_NumpyFoundation.ipynb) behavior.

In [53]:
x = torch.randn(2,3)
print (x)
print (x.shape)

tensor([[ 0.5548, -0.0845,  0.5903],
        [-1.0032, -1.7873,  0.0538]])
torch.Size([2, 3])


In [54]:
## Concatenation

y = torch.cat([x,x], dim = 0) #concat on a specified dimension
print (y)
print (y.shape)

tensor([[ 0.5548, -0.0845,  0.5903],
        [-1.0032, -1.7873,  0.0538],
        [ 0.5548, -0.0845,  0.5903],
        [-1.0032, -1.7873,  0.0538]])
torch.Size([4, 3])


In [55]:
# Stacking 

z = torch.stack([x,x], dim = 0) # stack on new dimension
print (z)
print (z.shape)

tensor([[[ 0.5548, -0.0845,  0.5903],
         [-1.0032, -1.7873,  0.0538]],

        [[ 0.5548, -0.0845,  0.5903],
         [-1.0032, -1.7873,  0.0538]]])
torch.Size([2, 2, 3])


### Gradients

We can determine gradients (rate of change) of our tensors with respect to their constituents using gradient bookkeeping. The gradient is a vector that points in the direction of greatest increase of a function. We'll be using gradients in the next lesson to determine how to change our weights to affect a particular objective function (ex. loss).

Imagine you want to teach a computer how to recognize handwritten digits. To do this, you create a neural network that takes in images of digits as input, and produces a prediction of what digit is in the image as output. When you train the network, you show it many examples of handwritten digits along with the correct answer, and you adjust the weights of the network so that it produces the correct answer for each example.

However, adjusting the weights manually would be very difficult because there are so many of them, and it would be very time-consuming. Instead, you use an optimization algorithm to automatically adjust the weights for you. The most common optimization algorithm used in deep learning is called stochastic gradient descent (SGD).

SGD works by adjusting the weights in the direction of the steepest descent of the loss function. The loss function measures how well the network is doing at recognizing digits. During training, the network produces a prediction for each image, and the loss function compares the prediction to the correct answer. The goal of training is to minimize the loss function, which means that the network is doing a good job at recognizing digits.

To adjust the weights using SGD, we need to compute the gradients of the loss function with respect to each weight in the network. These gradients tell us how much each weight needs to be adjusted to reduce the loss function. PyTorch makes it easy to compute gradients automatically using the `backward()` method. When you call `backward()` on the output of the network, PyTorch computes the gradients of the output with respect to all the weights in the network.

Once we have the gradients, we can use them to adjust the weights using SGD. SGD multiplies the gradients by a small number called the learning rate, and subtracts the result from the weights. This adjusts the weights in the direction that reduces the loss function.

Over time, as the weights are adjusted, the network gets better at recognizing digits. The gradients tell us how much to adjust the weights, and SGD does the adjusting automatically, so we don't have to do it manually.

In summary, gradients are a key component of training neural networks. They tell us how much to adjust the weights of the network to produce a better output, and they are computed during the backpropagation proces

In [56]:
# Tensors with gradient bookkeeping

x = torch.rand(3, 4, requires_grad=True)
y = 3*x + 2
z = y.mean() #calculate the mean of y
z.backward() # z has to be scalar
print(f"x: \n{x}")
print(f"x.grad: \n{x.grad}")
print (y)


x: 
tensor([[0.7379, 0.0846, 0.4245, 0.9778],
        [0.6800, 0.3151, 0.3911, 0.8943],
        [0.6889, 0.8389, 0.1780, 0.6442]], requires_grad=True)
x.grad: 
tensor([[0.2500, 0.2500, 0.2500, 0.2500],
        [0.2500, 0.2500, 0.2500, 0.2500],
        [0.2500, 0.2500, 0.2500, 0.2500]])
tensor([[4.2138, 2.2539, 3.2734, 4.9334],
        [4.0399, 2.9453, 3.1733, 4.6829],
        [4.0667, 4.5168, 2.5341, 3.9325]], grad_fn=<AddBackward0>)


In PyTorch, we can define tensors with a special attribute called `requires_grad`. When we set this attribute to `True`, PyTorch keeps track of all the operations that are performed on that tensor. The operations are recorded in a computational graph, which is then used to compute gradients of a function with respect to that tensor using backpropagation.

In the given code snippet, we define a random tensor `x` of size `(3, 4)` and set its `requires_grad` attribute to `True`. We then perform some operations on `x` and define a new tensor `y` which is a linear transformation of `x`. We then define a scalar tensor `z` which is the mean of `y`. Finally, we call the `backward()` method on z to compute the gradients of `z` with respect to `x`. Since `z` is a scalar, we don't need to specify any arguments to the `backward()` method.

After calling `backward()`, the gradients of `z` with respect to `x` are computed and stored in the grad attribute of `x`. We print the value of `x` and `x.grad` to see the values of `x` and its gradients.

### CUDA
CUDA (Compute Unified Device Architecture) is a parallel computing platform and application programming interface (API) developed by NVIDIA for use with its graphics processing units (GPUs)

In PyTorch, CUDA can be used to accelerate training and inference of deep learning models by offloading computations to the GPU. PyTorch provides an interface to easily move tensors to and from the GPU, and to execute operations on the GPU using CUDA. This can significantly speed up the computation time for large neural networks and large datasets.

In [57]:
# Is CUDA available?
print (torch.cuda.is_available())


False


If False (CUDA is not available), let's change that by following these steps: Go to Runtime > Change runtime type > Change Hardware accelerator to GPU > Click Save

In [58]:
# Set device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print (device)


cpu


In [59]:
x = torch.rand(2,3)
print (x.is_cuda)
x = torch.rand(2,3).to(device) # Tensor is stored on the GPU
print (x.is_cuda)


False
False


I am going to continue to use my CPU :-). Going to start a project of recognising handwritting